apps/fhir/server/authentication.py

```diff
@@ -8,10 +9,111 @@
+        "hash_lookup_type": hash_lookup_type,
+        "hash_lookup_mesg": hash_lookup_mesg,
+    }))
+
+
+def match_backend_patient_identifier(mbi_hash, hicn_hash):
+    '''
+    Matches a patient identifier via the backend FHIR server
+    using an MBI or HICN hash.
+
+    Summary:
+        1. mbi_hash is used for the initial lookup.
+        2. If there is a mbi lookup issue, the hicn_hash is used next.
+        3. If there is a hicn lookup issue, exceptions are raised.
+        4. A NotFound exception is raised if no match was found.
+
+    Returns:
+        fhir_id = Matched patient identifier.
+        backend_data = Passed thru to caller. Utilized in TestAuthentication.
+        hash_lookup_type = The type used for the successful lookup (M or H).
+
+    Raises:
+        UpstreamServerException: If hicn_hash search found duplicates.
+        NotFound: If both searches did not match a fhir_id.
+    '''
     auth_state = FhirServerAuth(None)
     certs = (auth_state['cert_file'], auth_state['key_file'])
-    # URL for patient ID.
+
+    # 1. mbi_hash is used for the initial lookup.
+    hash_lookup_type = "M"
+    hash_lookup_mesg = None
+
+    if mbi_hash is not None:
+        # URL for patient ID by mbi_hash.
+        url = get_resourcerouter().fhir_url + \
+            "Patient/?identifier=https%3A%2F%2Fbluebutton.cms.gov" + \
+            "%2Fresources%2Fidentifier%2Fmbi-hash%7C" + \
+            mbi_hash + \
+            "&_format=application%2Fjson%2Bfhir"
+        response = requests.get(url, cert=certs, verify=False)
+        response.raise_for_status()
+        backend_data = response.json()
+
+        # Check resource bundle total > 1
+        if backend_data.get('total', 0) > 1:
+            hash_lookup_mesg = "Duplicate beneficiaries found in Patient resource bundle total"
+
+        # Check number of resource entries > 1
+        if (
+            'entry' in backend_data
+            and len(backend_data['entry']) > 1
+            and hash_lookup_mesg is not None
+        ):
+            hash_lookup_mesg = "Duplicate beneficiaries found in Patient resource bundle entry"
+
+        # If resource bundle has an entry and total == 1, return match results
+        if (
+            'entry' in backend_data
+            and backend_data['total'] == 1
+            and len(backend_data['entry']) == 1
+            and hash_lookup_mesg is None
+        ):
+            fhir_id = backend_data['entry'][0]['resource']['id']
+            # Log for FhirIDFound type
+            hash_lookup_mesg = "FOUND beneficiary via mbi_hash"
+            log_fhir_id_is_found(fhir_id, mbi_hash, hicn_hash,
+                                 hash_lookup_type, hash_lookup_mesg)
+            return fhir_id, backend_data, hash_lookup_type
+
+    # Log for mbi FhirIDNotFound type
+    hash_lookup_mesg = "FHIR ID NOT FOUND for MBI hash lookup"
+    log_fhir_id_not_found(mbi_hash, hicn_hash,
+                          hash_lookup_type, hash_lookup_mesg)
```

Comment:
> line 80 - 90: fhir lookup using mbi, duplicate checked, if found duplicate, hash_lookup_mesg set to "Duplicate found ......", but not seen that message is logged down the process flow.
> 
> Also question:
> when a mbi lookup returned multiple benes, is it treated as a WARNING where warning logged and process proceed to hicn lookup, or an error where audit event logged and process stopped here and auth failed?

@JFU-GIT Good catch! My thoughts were that the message would be included in the fall through here; however, it is being overwritten.  I'll add a check for if `hash_lookup_mesg is None` here, so that one of those duplicate messages is logged.
I'm also thinking about changing the naming of the logging functions to be more readable. For example, `log_fhir_id_not_found` to `log_fhir_id_not_matched`.  Will keep the `"type": "FhirIDFound"`, since it is already logging this way.

Reply:
looks good